src/auth/index.ts

```diff
@@ -0,0 +1,683 @@
+import Context from '../Context';
+import authGrant from '../db/models/authGrant';
+import { default as authGrantee } from '../db/models/authGrantee';
+import authInvite from '../db/models/authInvite';
+import { AuthTargetId, default as authTarget } from '../db/models/authTarget';
+import { default as authToken } from '../db/models/authToken';
+import governingEntity from '../db/models/governingEntity';
+import operationCluster from '../db/models/operationCluster';
+import {
+  default as participant,
+  ParticipantId,
+} from '../db/models/participant';
+import participantRole from '../db/models/participantRole';
+import role from '../db/models/role';
+import rolePermittedAction from '../db/models/rolePermittedAction';
+import { organizeObjectsByUniqueProperty } from '../db/util';
+import { createDeferredFetcher } from '../db/util/deferred';
+import { InstanceOfModel } from '../db/util/types';
+import {
+  createRootContext,
+  createRootContextFromRequest,
+  LogContext,
+} from '../logging/logging';
+import { isDefined } from '../util';
+import { createBrandedValue } from '../util/types';
+import * as hid from './hid';
+import * as perms from './permissions';
+import { PermittedActionIdString } from './role.data';
+import {
+  calculatePermissionsFromRolesGrant,
+  filterValidRoleStrings,
+  RolesGrant,
+} from './roles';
+
+const randomBytes = promisify(crypto.randomBytes);
+
+export const P = perms.AUTH_PERMISSIONS;
+
+type ParticipantInstance = InstanceOfModel<ReturnType<typeof participant>>;
+type AuthGranteeInstance = InstanceOfModel<ReturnType<typeof authGrantee>>;
+type AuthTargetInstance = InstanceOfModel<ReturnType<typeof authTarget>>;
+type AuthTokenInstance = InstanceOfModel<ReturnType<typeof authToken>>;
+
+/**
+ * Extend the existing permitted actions with new actions.
+ *
+ * Right now, lists of permitted actions for each role are stored in
+ * the database using the tables permittedAction and rolePermittedAction.
+ *
+ * TODO: These tables should be removed,
+ * and the mapping should be stored in code instead.
+ *
+ * In the meantime, for new code requiring new actions,
+ * they can be added to this file and avoid the need for a migration or
+ * modifying roles in any way.
+ */
+type ExtendedPermittedAction = PermittedActionIdString | perms.GlobalPermission;
+
+type LegacySyncMethod = (param: {
+  actor: ParticipantId;
+  log: LogContext;
+  target: AuthTargetInstance;
+  grantee: AuthGranteeInstance;
+}) => Promise<void>;
+
+/**
+ * Mapping from role name to additional actions to grant in addition to those
+ * in the database.
+ *
+ * These additional permissions apply regardless as to whether or not a
+ * participantRole is tied to a specific object or not.
+ *
+ * @deprecated
+ */
+const ADDITIONAL_ACTIONS: {
+  [id: string]: ExtendedPermittedAction[] | undefined;
+} = {
+  hpcadmin: Object.values(perms.AUTH_PERMISSIONS.global),
+  planlead: ['canCreateOrganizations', 'viewCategories'],
+};
```

Comment:
for the library and the new API, it should only use / know about the new roles and actions